### PR TITLE
Guard against attempting to creating out-of-range spans in macOS File…

### PR DESCRIPTION
**Pull request by alexischr**
…systemWatcher

Fixes https://github.com/dotnet/corefx/issues/40054

### CODE REVIEW

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 1]
```diff
@@ -422,7 +422,9 @@ internal unsafe void Start()
                         // The base FileSystemWatcher does a match check against the relative path before combining with
                         // the root dir; however, null is special cased to signify the root dir, so check if we should use that.
                         ReadOnlySpan<char> relativePath = ReadOnlySpan<char>.Empty;
-                        if (!path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase))
+                        if (!path.Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase)
+                            && path.Length >= _fullDirectory.Length
+                            && _fullDirectory.AsSpan().Equals(path.Slice(0, _fullDirectory.Length), StringComparison.OrdinalIgnoreCase))
```
Comment (scalablecory) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: You can use `path.StartsWith(fullDirectory, StringComparison.OrdinalIgnoreCase)` for this.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 2]
```diff
@@ -467,8 +469,17 @@ internal unsafe void Start()
                             {
                                 // Remove the base directory prefix and add the paired event to the list of
                                 // events to skip and notify the user of the rename
-                                ReadOnlySpan<char> newPathRelativeName = events[pairedId].Span.Slice(_fullDirectory.Length);
-                                watcher.NotifyRenameEventArgs(WatcherChangeTypes.Renamed, newPathRelativeName, relativePath);
+                                if (events[pairedId].Span.Length >= _fullDirectory.Length
+                                    && _fullDirectory.AsSpan().Equals(events[pairedId].Span.Slice(0, _fullDirectory.Length), StringComparison.OrdinalIgnoreCase)
```
Comment (scalablecory) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: And `events[pairedId].Span.StartsWith(_fullDirectory, StringComparison.OrdinalIgnoreCase)`.

**Comment by danmosemsft** [comment]
@alexischr thanks for the submission. Do you plan to address feedback.?

**Comment by stephentoub** [comment]
Replaced by https://github.com/dotnet/corefx/pull/41957

**Pull request closed without merging**